pkg/activator/activator.go

```diff
@@ -0,0 +1,272 @@
+	"github.com/cortexlabs/cortex/pkg/workloads"
+	"go.uber.org/zap"
+	istionetworkingclient "istio.io/client-go/pkg/clientset/versioned/typed/networking/v1beta1"
+	kapps "k8s.io/api/apps/v1"
+	kmeta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+type ctxValue string
+
+const APINameCtxKey ctxValue = "apiName"
+
+type Activator interface {
+	Try(ctx context.Context, fn func() error) error
+}
+
+type activator struct {
+	activatorsMux     sync.RWMutex
+	trackersMux       sync.RWMutex
+	autoscalerClient  autoscaler.Client
+	apiActivators     map[string]*apiActivator
+	readinessTrackers map[string]*readinessTracker
+	istioClient       istionetworkingclient.VirtualServiceInterface
+	logger            *zap.SugaredLogger
+}
+
+func New(
+	istioClient istionetworkingclient.VirtualServiceInterface,
+	deploymentInformer cache.SharedIndexInformer,
+	virtualServiceInformer cache.SharedIndexInformer,
+	autoscalerClient autoscaler.Client,
+	logger *zap.SugaredLogger,
+) Activator {
+	log := logger.With(zap.String("apiKind", userconfig.RealtimeAPIKind.String()))
+
+	act := &activator{
+		apiActivators:     make(map[string]*apiActivator),
+		readinessTrackers: make(map[string]*readinessTracker),
+		istioClient:       istioClient,
+		logger:            log,
+		autoscalerClient:  autoscalerClient,
+	}
+
+	virtualServiceInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    act.addAPI,
+			UpdateFunc: act.updateAPI,
+			DeleteFunc: act.removeAPI,
+		},
+	)
+
+	deploymentInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: act.updateReadinessTracker,
+			UpdateFunc: func(_, newObj interface{}) {
+				act.updateReadinessTracker(newObj)
+			},
+			DeleteFunc: act.removeReadinessTracker,
+		},
+	)
+
+	return act
+}
+
+func (a *activator) Try(ctx context.Context, fn func() error) error {
+	apiNameValue := ctx.Value(APINameCtxKey)
+	apiName, ok := apiNameValue.(string)
+	if !ok || apiName == "" {
+		return errors.ErrorUnexpected("failed to get the api name from context")
+	}
+
+	act, err := a.getOrCreateAPIActivator(ctx, apiName)
+	if err != nil {
+		return err
+	}
+
+	tracker := a.getOrCreateReadinessTracker(apiName)
+
+	if act.inFlight() == 0 {
+		go a.awakenAPI(apiName)
+	}
+
+	return act.try(ctx, fn, tracker)
+}
+
+func (a *activator) getOrCreateAPIActivator(ctx context.Context, apiName string) (*apiActivator, error) {
+	a.activatorsMux.RLock()
+	act, ok := a.apiActivators[apiName]
+	a.activatorsMux.RUnlock()
+	if ok {
+		return act, nil
+	}
+
+	vs, err := a.istioClient.Get(ctx, workloads.K8sName(apiName), kmeta.GetOptions{})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	maxQueueLength, maxConcurrency, err := concurrencyFromAnnotations(vs.Annotations)
+	if err != nil {
+		return nil, err
+	}
+
+	apiAct := newAPIActivator(maxQueueLength, maxConcurrency)
+
+	a.activatorsMux.Lock()
+	a.apiActivators[apiName] = apiAct
+	a.activatorsMux.Unlock()
```

Comment:
It might be better to place the lock around this entire function rather than just the beginning and end, to prevent multiple threads from sneaking in (like we discussed for `(a *Autoscaler) Awaken()`